File: benchmarks/js_src/fasta.js
// Fasta benchmark — Node.js (V8) version
// Random DNA sequence generation

const IM = 139968;
const IA = 3877;
const IC = 29573;

let seed = 42;
function uint32_rand() {
    return (seed = (seed * IA + IC) % IM);
}

const alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

function repeat_fasta(seq, n) {
    const len = seq.length;
    let pos = 0;
    while (n > 0) {
        const line = Math.min(n, 60);
        let buf = '';
        for (let i = 0; i < line; i++) {
            buf += seq[pos++];
            if (pos >= len) pos = 0;
        }
        console.log(buf);
        n -= line;
    }
}

function random_fasta(symb, probs, n) {
    const cum = [];
    let sum = 0;
    for (let i = 0; i < symb.length; i++) {
        sum += probs[i];
        cum[i] = sum;
    }
    const imF = IM;

    while (n > 0) {
        const line = Math.min(n, 60);
        let buf = '';
        for (let i = 0; i < line; i++) {
            const r = uint32_rand();
            let j = 0;
            while (j < symb.length - 1 && r >= cum[j] * imF) j++;
            buf += symb[j];
        }
        console.log(buf);
        n -= line;
    }
}

const n = parseInt(process.argv[2] || '25000', 10);

console.log('>ONE Homo sapiens alu');
repeat_fasta(alu, n * 2);

console.log('>TWO IUB ambiguity codes');
const iub = "acgtBDHKMNRSVWY";
const iub_p = [0.27, 0.12, 0.12, 0.27, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02];
random_fasta(iub, iub_p, n * 3);

console.log('>THREE Homo sapiens frequency');
const hs = "acgt";
const hs_p = [0.3029549426680, 0.1979883004921, 0.1975473066391, 0.3015094502008];
random_fasta(hs, hs_p, n * 5);

File: benchmarks/py_src/fasta.py
#!/usr/bin/env python3
# Fasta benchmark — Python version
import sys

IM = 139968
IA = 3877
IC = 29573

seed = 42
def uint32_rand():
    global seed
    seed = (seed * IA + IC) % IM
    return seed

def repeat_fasta(seq, n):
    length = len(seq)
    pos = 0
    out_lines = []
    while n > 0:
        line = min(n, 60)
        buf = []
        for _ in range(line):
            buf.append(seq[pos])
            pos += 1
            if pos >= length:
                pos = 0
        out_lines.append(''.join(buf))
        n -= line
    return out_lines

def random_fasta(symb, probs, n):
    # Cumulative probabilities
    cum = []
    s = 0.0
    for p in probs:
        s += p
        cum.append(s)
    ns = len(symb)
    out_lines = []
    while n > 0:
        line = min(n, 60)
        buf = []
        for _ in range(line):
            r = uint32_rand()
            for j in range(ns):
                if r < cum[j] * IM:
                    buf.append(symb[j])
                    break
        out_lines.append(''.join(buf))
        n -= line
    return out_lines

def run(n):
    lines = []
    
    alu = ("GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
           "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
           "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
           "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
           "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
           "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
           "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA")
    
    lines.append(">ONE Homo sapiens alu")
    lines.extend(repeat_fasta(alu, n * 2))
    
    iub = "acgtBDHKMNRSVWY"
    iub_p = [0.27, 0.12, 0.12, 0.27, 0.02, 0.02, 0.02, 0.02,
             0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02]
    
    lines.append(">TWO IUB ambiguity codes")
    lines.extend(random_fasta(iub, iub_p, n * 3))
    
    hs = "acgt"
    hs_p = [0.3029549426680, 0.1979883004921, 0.1975473066391, 0.3015094502008]
    
    lines.append(">THREE Homo sapiens frequency")
    lines.extend(random_fasta(hs, hs_p, n * 5))
    
    sys.stdout.write("\n".join(lines) + "\n")

if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 25000
    run(n)
